crates/nu-plugin/src/plugin/interface/engine.rs

```diff
@@ -797,6 +797,37 @@ impl EngineInterface {
         self.write(PluginOutput::CallResponse(self.context()?, response))?;
         self.flush()
     }
+
+    /// Change whether the plugin process is in the foreground. This is necessary for terminal UI
+    /// apps to work properly, so that they can receive input and signals from the terminal.
+    ///
+    /// Plugins always start in the background (`false`). This should be reset back to false when
+    /// releasing control of the terminal again, otherwise the plugin may be terminated by signals
+    /// such as the interrupt signal (ctrl-c) erroneously.
+    ///
+    /// The exact implementation is operating system-specific. On Unix, this sets the process group
+    /// ID of the plugin to that of `nu` itself if `foreground` is `true`, and leaves the group if
+    /// set to `false`.
+    pub fn set_foreground(&self, foreground: bool) -> Result<(), ShellError> {
+        #[cfg(unix)]
+        {
+            use nix::unistd::Pid;
+            let new_pgid = if foreground {
+                Pid::parent()
+            } else {
+                Pid::from_raw(0)
+            };
+            nix::unistd::setpgid(Pid::this(), new_pgid).map_err(|err| ShellError::IOError {
```

Comment:
@IanManske thanks - you're absolutely right, `^echo foo | nu_plugin_explore` totally does cause problems, and I can see what's going on in `foreground.rs`. it sounds like probably rather than the plugin doing this, it should have the engine set the process group for the plugin based on the same rules